File: leetcode/java/BinaryTreeZigZagLevelOrderTraversal.java
import java.util.*;

public class BinaryTreeZigZagLevelOrderTraversal {

	public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        
		List<List<Integer>> ans = new ArrayList<List<Integer>>();
		if (root == null)
			return ans;

		boolean reverse = false;
		List<TreeNode> nextLevel = new ArrayList<TreeNode>();
		List<TreeNode> thisLevel = new ArrayList<TreeNode>();
		thisLevel.add(root);
		while (!thisLevel.isEmpty()) {
			
			List<Integer> l = new ArrayList<Integer>();
			for (TreeNode n : thisLevel) {
				l.add(n.val);
				if (n.left != null)
					nextLevel.add(n.left);
				if (n.right != null)
					nextLevel.add(n.right);
			}
			if (reverse)
				Collections.reverse(l);
			ans.add(l);
			thisLevel = nextLevel;
			nextLevel = new ArrayList<TreeNode>();
			reverse = !reverse;
		}
		return ans;
	}

	public static void main(String[] args) {
		
	}
}

File: leetcode/cpp/binaryTreeZigzagLevelOrderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> result;
        vector<TreeNode*> currentLevel;
        vector<TreeNode*> nextLevel;
        if (root) currentLevel.push_back(root);
        bool isReversed = false;
        while (currentLevel.size()) {
            vector<int> v;
            if (isReversed) {
                for (auto it=currentLevel.rbegin(); it != currentLevel.rend(); it++)
                    v.push_back((*it)->val);
            } else {
                for (TreeNode* it: currentLevel)
                    v.push_back(it->val);
            }
            result.push_back(v);
            for (TreeNode* it: currentLevel) {
                if (it->left) nextLevel.push_back(it->left);
                if (it->right) nextLevel.push_back(it->right);
            }
            swap(currentLevel, nextLevel);
            nextLevel.clear();
            isReversed = !isReversed;
        }
        return result;
    }
};